File: src/main/java/mobi/hsz/idea/gitignore/util/MatcherUtil.java
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package mobi.hsz.idea.gitignore.util;

import com.intellij.util.containers.IntObjectCache;
import org.apache.commons.lang.builder.HashCodeBuilder;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Pattern;

/**
 * Util class to speed up and limit regex operation on the files paths.
 *
 * @author Jakub Chrzanowski <jakub@hsz.mobi>
 * @since 1.3.1
 */
public class MatcherUtil {
    /** Stores calculated matching results. */
    private final IntObjectCache<Boolean> cache = new IntObjectCache<>();

    /**
     * Extracts alphanumeric parts from the regex pattern and checks if any of them is contained in the tested path.
     * Looking for the parts speed ups the matching and prevents from running whole regex on the string.
     *
     * @param pattern to explode
     * @param path    to check
     * @return path matches the pattern
     */
    public boolean match(@Nullable Pattern pattern, @Nullable String path) {
        if (pattern == null || path == null) {
            return false;
        }

        synchronized (cache) {
            int hashCode = new HashCodeBuilder().append(pattern).append(path).toHashCode();

            if (!cache.containsKey(hashCode)) {
                final String[] parts = getParts(pattern);
                boolean result = false;

                if (parts.length == 0 || matchAllParts(parts, path)) {
                    try {
                        result = pattern.matcher(path).find();
                    } catch (StringIndexOutOfBoundsException ignored) {
                    }
                }

                cache.put(hashCode, result);
                return result;
            }

            return cache.get(hashCode);
        }
    }

    /**
     * Checks if given path contains all of the path parts.
     *
     * @param parts that should be contained in path
     * @param path  to check
     * @return path contains all parts
     */
    public static boolean matchAllParts(@Nullable String[] parts, @Nullable String path) {
        if (parts == null || path == null) {
            return false;
        }

        int index = -1;
        for (String part : parts) {
            index = path.indexOf(part, index);
            if (index == -1) {
                return false;
            }
        }

        return true;
    }

    /**
     * Checks if given path contains any of the path parts.
     *
     * @param parts that should be contained in path
     * @param path  to check
     * @return path contains any of the parts
     */
    public static boolean matchAnyPart(@Nullable String[] parts, @Nullable String path) {
        if (parts == null || path == null) {
            return false;
        }

        for (String part : parts) {
            if (path.contains(part)) {
                return true;
            }
        }

        return false;
    }

    /**
     * Extracts alphanumeric parts from  {@link Pattern}.
     *
     * @param pattern to handle
     * @return extracted parts
     */
    @NotNull
    public static String[] getParts(@Nullable Pattern pattern) {
        if (pattern == null) {
            return new String[0];
        }

        final List<String> parts = new ArrayList<>();
        final String sPattern = pattern.toString();

        StringBuilder part = new StringBuilder();
        boolean inSquare = false;
        for (int i = 0; i < sPattern.length(); i++) {
            char ch = sPattern.charAt(i);
            if (!inSquare && Character.isLetterOrDigit(ch)) {
                part.append(sPattern.charAt(i));
            } else if (part.length() > 0) {
                parts.add(part.toString());
                part = new StringBuilder();
            }

            inSquare = ch != ']' && ((ch == '[') || inSquare);
        }

        return parts.toArray(new String[0]);
    }
}

File: src/main/kotlin/mobi/hsz/idea/gitignore/util/MatcherUtil.kt
// Copyright 2000-2021 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package mobi.hsz.idea.gitignore.util

import com.intellij.openapi.project.Project
import com.intellij.openapi.roots.ProjectRootManager
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.search.FilenameIndex
import com.intellij.psi.search.GlobalSearchScope
import java.util.ArrayList
import java.util.regex.Pattern

/**
 * Util class to speed up and limit regex operation on the files paths.
 */
class MatcherUtil private constructor() {

    companion object {
        /**
         * Checks if given path contains all of the path parts.
         *
         * @param parts that should be contained in path
         * @param path  to check
         * @return path contains all parts
         */
        @Suppress("ReturnCount")
        fun matchAllParts(parts: Array<String?>?, path: String?): Boolean {
            var index = -1
            parts?.filterNotNull()?.forEach {
                index = path?.indexOf(it, index) ?: -1
                if (index == -1) {
                    return false
                }
            } ?: return false
            return true
        }

        /**
         * Checks if given path contains any of the path parts.
         *
         * @param parts that should be contained in path
         * @param path  to check
         * @return path contains any of the parts
         */
        fun matchAnyPart(parts: Array<String?>?, path: String?) = parts?.filterNotNull()?.any { path?.contains(it) ?: false } ?: false

        /**
         * Extracts alphanumeric parts from  [Pattern].
         *
         * @param pattern to handle
         * @return extracted parts
         */
        fun getParts(pattern: Pattern?): Array<String?> {
            if (pattern == null) {
                return arrayOfNulls(0)
            }
            val parts: MutableList<String?> = ArrayList()
            val sPattern = pattern.toString()
            var part = StringBuilder()
            var inSquare = false
            for (i in sPattern.indices) {
                val ch = sPattern[i]
                if (!inSquare && Character.isLetterOrDigit(ch)) {
                    part.append(sPattern[i])
                } else if (part.isNotEmpty()) {
                    parts.add(part.toString())
                    part = StringBuilder()
                }
                inSquare = ch != ']' && (ch == '[' || inSquare)
            }
            return parts.toTypedArray()
        }

        /**
         * Finds [VirtualFile] instances for the specific [Pattern] and caches them.
         *
         * @param project current project
         * @param pattern to handle
         * @return matched files list
         */
        fun getFilesForPattern(project: Project, pattern: Pattern): Collection<VirtualFile> {
            val parts = getParts(pattern).ifEmpty { return emptyList() }
            val projectFileIndex = ProjectRootManager.getInstance(project).fileIndex
            val scope = GlobalSearchScope.projectScope(project)
            val files = mutableSetOf<VirtualFile>()

            projectFileIndex.iterateContent {
                if (matchAnyPart(parts, it.name)) {
                    FilenameIndex.getVirtualFilesByName(project, it.name, scope).forEach { file ->
                        if (file.isValid && matchAllParts(parts, file.path)) {
                            files.add(file)
                        }
                    }
                }
                true
            }

            return files
        }
    }
}
